Documentation: Document CMake build options (#7112)
SerenityOS/serenity: Documentation/BuildInstructions.md

### CMake build options

There are some optional features that can be enabled during compilation that are intended to help with specific types of development work or introduce experimental features. Currently, the following build options are available:
- `ENABLE_ADDRESS_SANITIZER` and `ENABLE_KERNEL_ADDRESS_SANITIZER`: builds in runtime checks for memory corruption bugs (like buffer overflows and memory leaks) in Lagom test cases and the kernel, respectively.
- `ENABLE_MEMORY_SANITIZER`: enables runtime checks for uninitialized memory accesses in Lagom test cases.
- `ENABLE_UNDEFINED_SANITIZER`: builds in runtime checks for [undefined behavior](https://en.wikipedia.org/wiki/Undefined_behavior) (like null pointer dereferences and signed integer overflows) in Lagom test cases.
- `ENABLE_FUZZER_SANITIZER`: builds [fuzzers](https://en.wikipedia.org/wiki/Fuzzing) for various parts of the system.
- `ENABLE_ALL_THE_DEBUG_MACROS`: used for checking whether debug code compiles on CI. This should not be set normally, as it clutters the console output and makes the system run very slowly. Instead, enable only the needed debug macros, as described below.

> - `ENABLE_COMPILETIME_FORMAT_CHECK`: checks for the validity of Python-style format string during compilation. Enabled by default.

Reviewer: ```suggestion
- `ENABLE_COMPILETIME_FORMAT_CHECK`: checks for the validity of `std::format`-style format string during compilation. Enabled by default.
```

It's indeed similar to Python's format strings, but what we have is essentially a clone of `std::format` syntax :)

> - `ENABLE_PCI_IDS_DOWNLOAD`: downloads the [`pci.ids` database](https://pci-ids.ucw.cz/) that contains information about PCI devices at build time. Enabled by default.

Reviewer: ```suggestion
- `ENABLE_PCI_IDS_DOWNLOAD`: downloads the [`pci.ids` database](https://pci-ids.ucw.cz/) that contains information about PCI devices at build time, if not already present. Enabled by default.
```

I think it's important to mention that this only happens once for every fresh build.

> - `BUILD_LAGOM`: builds [Lagom](../Meta/Lagom/ReadMe.md), which can be used for testing certain SerenityOS libraries on the host system.

Reviewer: ```suggestion
- `BUILD_LAGOM`: builds [Lagom](../Meta/Lagom/ReadMe.md), which makes various SerenityOS libraries and programs available on the host system.
```

- `PRECOMPILE_COMMON_HEADERS`: precompiles some common headers to speedup compilation.
- `ENABLE_KERNEL_LTO`: builds the kernel with link-time optimization.

> Many parts of Serenity can be set to print detailed debug information to the debug console. This is done via the `<component_name>_DEBUG` macros, which can also be enabled when building Serenity. They are listed in [this file](../Meta/CMake/all_the_debug_macros.cmake).

Reviewer: ```suggestion
Many parts of the SerenityOS codebase have debug functionality, mostly consisting of additional messages printed to the debug console. This is done via the `<component_name>_DEBUG` macros, which can be enabled individually at build time. They are listed in [this file](../Meta/CMake/all_the_debug_macros.cmake).
```

To toggle a build option, add it to the `cmake` command invocation with a `-D` prefix. To enable it, add `=ON` at the end, or add `=OFF` to disable it. The complete command should look similarly to this:

```console
$ cmake ../.. -G Ninja -DPROCESS_DEBUG=ON -DENABLE_PCI_IDS_DOWNLOAD=OFF
```

> For the changes to take effect, Serentiy needs to be recompiled and the disk image needs to be rebuild.

Reviewer: ```suggestion
For the changes to take effect, SerenityOS needs to be recompiled and the disk image needs to be rebuild.
```
